perception/alars/auv_detector/auv_detector/hook_estimator.py

```diff
@@ -0,0 +1,1020 @@
+
+from scipy.spatial.transform import Rotation
+import matplotlib.pyplot as plt
+import matplotlib
+matplotlib.use('Agg')
+from collections import deque
+import os
+import csv
+from itertools import groupby
+from operator import itemgetter
+from filterpy.kalman import ExtendedKalmanFilter
+
+from scipy.spatial.transform import Rotation
+from datetime import datetime
+
+
+class hook_estimator(Node):
+    def __init__(self):
+        super().__init__('k_nearest_neighbors')
+
+        self.subscription_camera = self.create_subscription(
+            Image,
+            '/Quadrotor/core/fpcamera/image',
+            self.listener_callback,            
+            10)
+        self.subscription_camera  # prevent unused variable warning
+
+        self.bridge = CvBridge()
+
+        self.cv_image = None
+        self.stamp = None
+        self.debug_imshow = True
+
+        self.hook_timer = self.create_timer(0.1, self.process_hook_estimation)
+
+        self.subscription_accel = self.create_subscription(
+            Imu,
+            '/Quadrotor/core/imu',
+            self.imu_callback,
+            10)
+
+        # Initialization (in __init__ or once)
+        self.est_Hook_pos_in_winch_frame = np.array([np.nan, np.nan, np.nan])
+        self.est_Hook_pos_in_base_link_frame = np.array([np.nan, np.nan, np.nan])
+        self.ekf_Hook_pos_in_winch_frame = np.array([np.nan, np.nan, np.nan])
+        self.ekf_Hook_pos_in_base_link_frame = np.array([np.nan, np.nan, np.nan])
+        self.est_Hook_pos_in_winch_frame_ekf = np.array([np.nan, np.nan, np.nan])
+        self.Hook_pos_in_base_link_frame_to_plot = deque(maxlen=2000)
+        self.est_Hook_pos_in_base_link_frame_to_plot = deque(maxlen=2000)
+        self.ekf_Hook_pos_in_base_link_frame_to_plot = deque(maxlen=2000)
+        self.est_hook_timestamps = deque(maxlen=2000)  # Store timestamps of hook detections
+        self.camera_sees_hook = deque(maxlen=2000)
+
+        self.plot_initialized = False
+        self.ekf_est_hook_map_pos_to_save = deque(maxlen=2000)  # Store EKF hook estimated pos
+        self.ekf_initialized = False
+        self.last_ekf_time = None
+
+        self.current_length = 5.0 # Length of the rope in meters, can be changed by winch control
+
+        # EKF setup
+        self.ekf = ExtendedKalmanFilter(dim_x=6, dim_z=3)  # State: [p, v]; Measured: [r]
+        self.ekf.x = np.zeros(6)
+        self.ekf.P *= 1e-1
+        self.ekf.R = np.eye(3) * 1e-1                       # Measurement noise (camera)
+        self.ekf.Q = np.eye(6) * 1e-2                       # Process noise
+        self.g_vec = np.array([0, 0, 0])  # Gravity vector in base_link frame
+
+        # Initialize IMU values
+        self.a_drone = None
+        self.smoothed_accel = None
+        self.a_drone_odom = None
+
+        # Initialize the tf buffer and listener
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(self.tf_buffer, self)
+
+        # Set frames
```

Comment:
_ALL_ of these frames should be coming from dji_msgs.msg.LINKS similar to topics. 
however, links can not be namespaced like topics, so your node needs to take `robot_name` as a rosparam (and the launch file that launches it).

example: https://github.com/smarc-project/smarc2/blob/d07b9ffea6c5b28e17e5665fcecc90a6250abca3/vehicles/hardware/dji/dji_captain/dji_captain/dji_captain.py#L106